### PR TITLE
Add ability to stop sync to target clusters

### PR DESCRIPTION
**Pull request by wanghaoran1988**

### CODE REVIEW

File: docs/annotations.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Annotations
+
+This document provides a list of annotations used by hive CRDs
+
+
+| Annotation| Description | 
+| ---------- | ----------- |
+| hive.openshift.io/syncset-pause | When the value is "true", hive will stop syncing everything to target cluster including resources defined in `syncset` object, and remote machineset.  |
```
Comment (LalatenduMohanty) [inline docs/annotations.md]: :+1:

File: docs/annotations.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Annotations
+
+This document provides a list of annotations used by hive CRDs
```
Comment (LalatenduMohanty) [inline docs/annotations.md]: s/hive/Hive/

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -39,6 +39,9 @@ const (
 
 	// CheckpointName is the name of the object in each namespace in which the namespace's backup information is stored.
 	CheckpointName = "hive"
+
+	// AnnotationSyncsetPause is a annotation used by clusterDeployment, if it's true, then we will disable syncing to a specific cluster
+	AnnotationSyncsetPause = "hive.openshift.io/syncset-pause"
```
Comment (LalatenduMohanty) [inline pkg/constants/constants.go]: s/AnnotationSyncsetPause/SyncsetPauseAnnotation/ . IMO `SyncsetPauseAnnotation` is a better name.

File: pkg/controller/remotemachineset/remotemachineset_controller.go [hunk 4]
```diff
@@ -128,6 +129,10 @@ func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconc
 		log.WithError(err).Error("error looking up cluster deployment")
 		return reconcile.Result{}, err
 	}
+	if cd.Annotations != nil && cd.Annotations[constants.AnnotationSyncsetPause] == "true" {
+		log.Debug("syncing to cluster is disabled")
```
Comment (LalatenduMohanty) [inline pkg/controller/remotemachineset/remotemachineset_controller.go]: I would prefer this to be a info message as this information should be available in the logs easily as we do not want users to forget about the annotation and then wondering why syncsets are not working.

File: pkg/controller/remotemachineset/remotemachineset_controller.go [hunk 5]
```diff
@@ -128,6 +129,10 @@ func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconc
 		log.WithError(err).Error("error looking up cluster deployment")
 		return reconcile.Result{}, err
 	}
+	if cd.Annotations != nil && cd.Annotations[constants.AnnotationSyncsetPause] == "true" {
```
Comment (LalatenduMohanty) [inline pkg/controller/remotemachineset/remotemachineset_controller.go]: Do we need to both `cd.Annotations != nil` and `cd.Annotations[constants.AnnotationSyncsetPause]`?

File: pkg/controller/remotemachineset/remotemachineset_controller.go (hunk 4 above)
Comment (dgoodwin) [reply pkg/controller/remotemachineset/remotemachineset_controller.go]: Lets actually go Warn for this, it's a one off customization that should not be done and forgotten.

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go [hunk 6]
```diff
@@ -198,6 +199,11 @@ func (r *ReconcileSyncSetInstance) Reconcile(request reconcile.Request) (reconci
 		return reconcile.Result{}, err
 	}
 
+	if cd.Annotations != nil && cd.Annotations[constants.AnnotationSyncsetPause] == "true" {
+		ssiLog.Debug("syncing to cluster is disabled")
```
Comment (dgoodwin) [inline pkg/controller/syncsetinstance/syncsetinstance_controller.go]: Warn here as well, I'd throw in "due to %s annotation" as well in both messages.

File: pkg/controller/remotemachineset/remotemachineset_controller.go (hunk 4 above)
Comment (LalatenduMohanty) [reply pkg/controller/remotemachineset/remotemachineset_controller.go]: Also lets mention in the message that `SyncsetPauseAnnotation is present, hence syncing to the cluster deployment is disabled`

File: pkg/controller/remotemachineset/remotemachineset_controller.go (hunk 5 above)
Comment (staebler) [inline pkg/controller/remotemachineset/remotemachineset_controller.go]: The check for `nil` is not necessary.

Comment (staebler) [inline pkg/controller/remotemachineset/remotemachineset_controller.go]: You probably want this to be the following for more flexibility.
```
if p, err := strconv.ParseBool(cd.Annotations; p && err != nil {
``

**Comment by wanghaoran1988** [comment]
Comments addressed, PTAL

File: pkg/controller/remotemachineset/remotemachineset_controller.go [hunk 7]
```diff
@@ -128,6 +129,10 @@ func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconc
 		log.WithError(err).Error("error looking up cluster deployment")
 		return reconcile.Result{}, err
 	}
+	if cd.Annotations[constants.SyncsetPauseAnnotation] == "true" {
+		log.Warn(constants.SyncsetPauseAnnotation, " is present, hence syncing to cluster is disabled")
```
Comment (dgoodwin) [inline pkg/controller/remotemachineset/remotemachineset_controller.go]: Usually would see this as log.Warnf("%s is present, hence ...", constants.SyncsetPauseAnnotation) but no idea if either form is better or why.

Comment (dgoodwin) [reply pkg/controller/remotemachineset/remotemachineset_controller.go]: Just a comment, not a necessary change.

File: pkg/controller/remotemachineset/remotemachineset_controller.go (hunk 5 above)
Comment (dgoodwin) [reply pkg/controller/remotemachineset/remotemachineset_controller.go]: @wanghaoran1988 This comment was probably missed, the above format would be a little more flexible and let you be less strict on the format of "true" in the annotation.

**Pull request merged**